File: tempFiles/browserJava/src/main/java/com/luciasoft/browserjava/collections/BinarySearchTree.java
package com.luciasoft.browserjava.collections;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Iterator;

class Node<T>
{
    Node(T data, Node<T> prnt)
    {
        this.data = data;
        this.prnt = prnt;
    }

    T data;
    Node<T> left;
    Node<T> rght;
    Node<T> prnt;
}

public class BinarySearchTree<T> implements Iterable<T>
{
    public BinarySearchTree(Comparator<T> comparator)
    {
        this.comparator = comparator;
    }

    Node<T> root;

    int count;

    Comparator<T> comparator;

    public int getCount()
    {
        return count;
    }

    public Comparator<T> getComparator()
    {
        return comparator;
    }

    public void add(T data)
    {
        if (root == null)
        {
            root = new Node(data, null);

            count++;

            return;
        }

        add(data, root);
    }

    void add(T data, Node<T> node)
    {
        int cmp = comparator.compare(data, node.data);

        if (cmp < 0)
        {
            if (node.left == null)
            {
                node.left = new Node(data, node);

                count++;

                return;
            }

            add(data, node.left);

            return;
        }

        if (cmp > 0)
        {
            if (node.rght == null)
            {
                node.rght = new Node(data, node);

                count++;

                return;
            }

            add(data, node.rght);
        }

        // cmp == 0
    }

    public T getData(T data)
    {
        if (root == null) return null;

        return getData(data, root);
    }

    T getData(T data, Node<T> node)
    {
        int cmp = comparator.compare(data, node.data);

        if (cmp < 0)
        {
            if (node.left == null) return null;

            return getData(data, node.left);
        }

        if (cmp > 0)
        {
            if (node.rght == null) return null;

            return getData(data, node.rght);
        }

        return node.data;
    }

    public void reset()
    {
        root = null;
        count = 0;
    }

    public void traverseInOrder(DoWithData<T> doWithData)
    {
        traverseInOrder(root, doWithData);
    }

    void traverseInOrder(Node<T> curr, DoWithData<T> doWithData)
    {
        if (curr == null) return;

        traverseInOrder(curr.left, doWithData);
        doWithData.doWithData(curr.data);
        traverseInOrder(curr.rght, doWithData);
    }

    public ArrayList<T> toList()
    {
        final ArrayList<T> list = new ArrayList<>();

        T[] one = null;

        traverseInOrder(new DoWithData<T>()
        {
            @Override
            public void doWithData(T data)
            {
                list.add(data);
            }
        });

        return list;
    }

    @Override
    public Iterator<T> iterator()
    {
        return new TreeIterator<>(this);
    }
}

class TreeIterator<T> implements Iterator<T>
{
    TreeIterator(BinarySearchTree<T> tree)
    {
        this.tree = tree;

        initialize();
    }

    Node<T> curr = null;

    void initialize()
    {
        curr = getLeftMost(tree.root);
    }

    @Override
    public boolean hasNext()
    {
        return curr != null;
    }

    @Override
    public T next()
    {
        if (curr == null) return null;

        T data = curr.data;

        curr = getNextNode(curr);

        return data;
    }

    BinarySearchTree<T> tree;

    static Node getLeftMost(Node node)
    {
        while (node.left != null) node = node.left;

        return node;
    }

    static Node getNextNode(Node node)
    {
        if (node.rght != null)
        {
            return getLeftMost(node.rght);
        }
        else
        {
            while (node.prnt != null && node == node.prnt.rght) node = node.prnt;

            return node.prnt;
        }
    }
}

interface DoWithData<T>
{
    void doWithData(T data);
}

File: collections/src/main/java/com/luciasoft/collections/BinarySearchTree.kt
package com.luciasoft.collections

class BSTofString(allowDupes: Boolean, ignoreCase: Boolean, special: Boolean) : BinarySearchTree<String>(allowDupes, StringComparator(ignoreCase, special))

open class BinarySearchTree<T>(private val allowDupes: Boolean, private val comparator: Comparator<T>) : Iterable<T>
{
    private class Node<T>(val data: T, val prnt: Node<T>?)
    {
        var left: Node<T>? = null
        var rght: Node<T>? = null
    }

    private var root: Node<T>? = null

    var count: Int = 0
    private set

    fun getSameAs(data: T): T?
    {
        return getSameAs(data, root)
    }

    private fun getSameAs(data: T, node: Node<T>?): T?
    {
        if (root == null) return null

        checkNotNull(node) {"node cannot be null."}

        val compare = comparator.compare(data, node.data)

        if (compare < 0) return getSameAs(data, node.left)
        if (compare > 0) return getSameAs(data, node.rght)

        return node.data
    }
    
    fun add(data: T)
    {
        add(data, root)
    }

    fun add(vararg data: T)
    {
        for (datum in data) add(datum)
    }

    fun add(randomize: Boolean, vararg data: T)
    {
        if (!randomize) for (datum in data) add(datum)
        else add(data.toList(), randomize)
    }

    fun add(data: Collection<T>, randomize: Boolean)
    {
        if (!randomize)
        {
            for (datum in data) add(datum)
        }
        else
        {
            val list = randomize(data);
            for (datum in list) add(datum)
        }
    }

    private fun add(data: T, node: Node<T>?)
    {
        if (root == null)
        {
            root = Node(data, null)
            count++
            return
        }

        checkNotNull(node) {"node cannot be null."}

        var compare = comparator.compare(data, node.data)
        if (allowDupes && compare == 0) compare = 1

        if (compare < 0)
        {
            if (node.left == null)
            {
                node.left = Node(data, node)
                count++
                return
            }

            add(data, node.left)
            return
        }

        if (compare > 0)
        {
            if (node.rght == null)
            {
                node.rght = Node(data, node)
                count++
                return
            }

            add(data, node.rght)
            return
        }
    }

    override fun iterator(): Iterator<T> = MyIterator(this)

    private class MyIterator<T>(private val tree: BinarySearchTree<T>) : Iterator<T>
    {
        private var current = getFirst()

        override fun hasNext(): Boolean = current != null

        override fun next(): T
        {
            val data = current!!.data
            current = getNext(current!!)
            return data
        }

        private fun getFirst(): Node<T>?
        {
            if (tree.root == null) return null;
            return getLeftMost(tree.root!!)
        }

        private fun getNext(node: Node<T>): Node<T>?
        {
            var n = node
            if (n.rght != null) return getLeftMost(n.rght!!);
            while (n.prnt != null && n == n.prnt!!.rght) n = n.prnt!!
            return n.prnt
        }
        
        private fun getLeftMost(node: Node<T>): Node<T>
        {
            var n = node
            while (n.left != null) n = n.left as Node<T>
            return n
        }
    }
}